### PR TITLE
Enhance rule metadata and suppression info in SARIF V2 errorlog

### PR DESCRIPTION
**Pull request by mavasani**
Two additional piece of data is now logged in SARIF v2 log:
1. Inside `Results` section for reported diagnostics: `Results` section contains an entry for each reported diagnostic. For each diagnostic instance that is suppressed with a pragma, SuppressMessageAttribute or via a DiagnosticSuppressor, we now report a new property `suppressionType` with one of these three values. This data helps analyze the preferred in-source suppression mechanisms in a code base.
2. Inside `Rules` metadata section for analyzer rules: `Rules` section contain an entry for each analyzer reported `DiagnosticDescriptor`. Prior to this PR, we only logged rule metadata, i.e. `DiagnosticDescriptor` data, for those rules that fired at least one diagnostic instance in the compilation, and the rule metadata had no suppression info, so there was no way to know which all rules were executed on the project and which were disabled for part or entirety of the compilation. We now report the `DiagnosticDescriptor` info for _all_ analyzers that were provided to the compilation, regardless of whether or not they executed on the entire compilation, part of the compilation or were disabled for the entire compilation. Additionally, if a rule had either a source suppression or was disabled for part or whole of the compilation via options, the rule metadata contains a special flag `isEverSuppressed = true` and an array `suppressionKinds` with either or both of the below suppression kinds:
   1. `inSource` suppression kind for one or more reported diagnostic(s) that were suppressed through pragma directive, SuppressMessageAttribute or a DiagnosticSuppressor.
   2. `external` suppression kind for diagnostic ID that is disabled either for the entire compilation (via global options such as /nowarn, ruleset, globalconfig, etc.) or for certain files or folders in the compilation (via editorconfig options).
  

Sample [errorlog.txt](https://github.com/dotnet/roslyn/files/9799752/errorlog.txt)

**TODO:**
- [x] Add unit tests

### CODE REVIEW

**Comment by jaredpar** [comment]
Could we add a doc file that captures much of the information in your PR message and include it in the PR? It's very valuable to know what these items mean and I think your PR has a great starting point for a doc. 

Doesn't need to be complex, just something we can point customers to in the future that explain what entries mean what. Can be roughly your PR description with a few formatting changes.

File: docs/compilers/Error Log Format.md
Comment (mavasani) [inline docs/compilers/Error Log Format.md:19]:
```suggestion
2. `runs` information: The core entry in the error log file is the `runs` section with a single run entry within it for the build. The run entry has 3 main parts:
```

File: docs/compilers/Error Log Format.md
Comment (mavasani) [inline docs/compilers/Error Log Format.md:108]:
```suggestion
6. `properties`: One or more custom properties associated with the descriptor. It includes the following:
```

**Comment by mavasani** [comment]
@jaredpar Thanks for the suggestion. I have updated the error log format doc with a high level overview of the SARIF v2 format error log contents.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for reviews.

**Comment by jaredpar** [comment]
@jjonescz, @cston, @RikkiGibson  PTAL

File: src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/SarifErrorLogger.cs:73]: >diagostics [](http://example.com/codeflow?start=60&length=10)

Typo

File: src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs:88]: >Suppression Id: {idAndJustification.Id}, Suppression Justification: {idAndJustification.Justification} [](http://example.com/codeflow?start=56&length=102)

Is this a JSON string? If not, are these names and values converted into JSON?

File: src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs:119]: >.OrderBy(d => d.Descriptor.Id) [](http://example.com/codeflow?start=79&length=30)

Is `OrderBy()` necessary, or will `_descriptors.Add()` store the items in a dictionary that ignores the order?

File: src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs:280]: >tags [](http://example.com/codeflow?start=56&length=4)

suppressionKinds?

File: src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs
Comment (cston) [inline src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs:379]: >KeyValuePair<int, (DiagnosticDescriptor Descriptor, bool HasAnyExternalSuppression)> [](http://example.com/codeflow?start=24&length=84)

Could this be simplified to `List<(int, DiagnosticDescriptor, bool)>`?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:283]: >_lazySuppressedDiagnosticIdsForUnsuppressedAnalyzers != null [](http://example.com/codeflow?start=29&length=60)

How do we know `_lazySuppressedDiagnosticIdsForUnsuppressedAnalyzers != null`? It looks like the field is only set by a parameter to `Initialize()` which is marked nullable.

File: src/Compilers/CSharp/Test/CommandLine/SarifV2ErrorLoggerTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/CommandLine/SarifV2ErrorLoggerTests.cs:439]: >Category2 [](http://example.com/codeflow?start=51&length=9)

Category1?

File: src/Compilers/CSharp/Test/CommandLine/SarifV2ErrorLoggerTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/CommandLine/SarifV2ErrorLoggerTests.cs:450]: >ID1 [](http://example.com/codeflow?start=18&length=3)

It looks like there is one suppression here. Are there diagnostics for ID2?

File: src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs
Comment (mavasani) [reply src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs:88]: No its just a ToString representation of the programmatic suppression info that is inserted in the the error log as a JSON value string. Do you feel there could be a better way to emit this data?

File: src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs
Comment (mavasani) [reply src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs:119]: `_descriptors.Add()` incrementally assigns an index to each added entry, and if we don't invoke `OrderBy` here, the indices will not be sorted by ID, and the rule entries in the error log will also be in a non-deterministic order.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs
Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:283]: This is the pattern used for most lazy backing fields and property pairs in this file. The property getter is asserting that it is accessed only on those code paths where this field is guaranteed to be non-null. In this case, the API `GetAllDescriptors` added in this PR is invoked from the command line compiler only when `/errorlog` switch is specified on the command line, which it turns passes `trackSuppressedDiagnosticIds: true` to the analyzer driver's `Initialize()` method and ensures the field is assigned a non-null value during initialization.

File: src/Compilers/CSharp/Test/CommandLine/SarifV2ErrorLoggerTests.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/CommandLine/SarifV2ErrorLoggerTests.cs:450]: ID2 is a project level no-location diagnostic that cannot be suppressed with editorconfig or pragma/SuppressMessAttribute. 
https://github.com/dotnet/roslyn/blob/bc641a6e58a5c15ac6b8052243c2b0dd97e65e1e/src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs#L71-L72
and
https://github.com/dotnet/roslyn/blob/bc641a6e58a5c15ac6b8052243c2b0dd97e65e1e/src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs#L41-L50

Below, the test invokes `GetExpectedV2ErrorLogRulesText` with a non-null `suppressionKinds1` and null `suppressionKinds2` to get the expected rules text with `ID1` suppressed and `ID2` unsuppressed. This helper method returns an expected entry for both ids, but suppression info added only for the ID with non-null `suppressionKinds` argument.

File: src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs
Comment (cston) [reply src/Compilers/Core/Portable/CommandLine/SarifV2ErrorLogger.cs:88]: > Do you feel there could be a better way to emit this data?

No, I just wanted to make sure I understood how it was represented, thanks.

**Review by cston — APPROVED** [review]

**Comment by mavasani** [comment]
@jjonescz @RikkiGibson for second review, thanks!

**Comment by mavasani** [comment]
@jjonescz @RikkiGibson @dotnet/roslyn-compiler for second review, thanks!

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler can I please get a second review? Thanks!

**Comment by mavasani** [comment]
Merging this in to unblock the security team from using the additional SARIF info. I'll address any additional feedback with a follow-up PR. Thanks!

**Pull request merged**
